### ISSUE TITLE
incorrect error on double import

### ISSUE THREAD

**Issue by lukeapage** [rollup/rollup#148]
I think this might be from the recent fix

```
// bar.js
export default function bar() {
}

// foo.js
export { default as bar } from "./bar";

// zoo.js
import * as foo from './foo';

function wah() {
  foo.bar();
}

wah();

// entry.js
import * as foo from './foo';
import './zoo';

export default {
  foo: foo
};
```

error is..

```
Error: Module "c:/git/grunt-rollup/test/fixtures/foo.js" doesn't export "bar"!
    at walk.leave (c:\git\grunt-rollup\node_modules\rollup\dist\rollup.js:1881:14)
    at visit (c:\git\grunt-rollup\node_modules\rollup\dist\rollup.js:1517:3)
    at visit (c:\git\grunt-rollup\node_modules\rollup\dist\rollup.js:1512:4)
    at visit (c:\git\grunt-rollup\node_modules\rollup\dist\rollup.js:1512:4)
    at visit (c:\git\grunt-rollup\node_modules\rollup\dist\rollup.js:1509:5)
    at visit (c:\git\grunt-rollup\node_modules\rollup\dist\rollup.js:1512:4)
    at walk (c:\git\grunt-rollup\node_modules\rollup\dist\rollup.js:1463:2)
    at Statement.analyse (c:\git\grunt-rollup\node_modules\rollup\dist\rollup.js:1808:4)
    at c:\git\grunt-rollup\node_modules\rollup\dist\rollup.js:2499:14
    at Array.forEach (native)
```

I can't see a problem with it though ? unless its the wrong error message and there is something else I am missing.

again stumbled on this whilst trying to reproduce another bug but it could be related so I'll wait on reporting that till this is fixed.

**Comment by lukeapage** [comment]
If I change entry.js to..
```
import './zoo';
```

then it all compiles fine...

oh and this is on master, not 1.16.4

**Comment by Victorystick** [comment]
It's a recent change designed so that we can better track when an undefined export is referenced. I'll investigate the issue today, and see what I can do.

**Comment by Victorystick** [comment]
The issue is caused by the order in which the modules are analysed. Due to how the modules depend on each other (and how they are loaded), `zoo` is analysed before `foo`, meaning that `foo`'s exports are undefined when `zoo` attempts to access them. This could be avoided by splitting the analysis into two stages:

1. Where all imports/exports are defined and bound.
2. Where all statements are analysed with respect to these bindings.

**Comment by Rich-Harris** [comment]
Yeah, I think namespace optimisation will need to happen at render time, not analysis time. Can it be treated like any other identifier replacement? (Notwithstanding nuance around determining which namespaces need to exist as blocks of getters.)

**Comment by Victorystick** [comment]
Yes. Having tried out the two phase analysis I suggested (and watched it fail miserably), I think I agree with you. It'd be nice to have an optimisation phase, after the main analysis is carried out.

**Comment by Rich-Harris** [comment]
This is a gross oversimplification, but isn't it mostly a case of moving [this logic](https://github.com/rollup/rollup/blob/master/src/Statement.js#L194-L279) down [here](https://github.com/rollup/rollup/blob/master/src/Statement.js#L552)?

**Comment by lukeapage** [comment]
this might fix some problems i have with namespaces not being optimised
(the problem i was trying to reproduce when i found this)

**Comment by Victorystick** [comment]
I'm not sure. I worry that by the time `replaceIdentifiers` is called, the namespaces that requires dynamic access have already been determined, and therefore cannot be optimised away at that stage.

However, I was able to quite easily extract the entire analysis into a separate pass just after all modules have been loaded. In `Bundle.js`, this would do the trick. All test pass, including the one I created for this issue.
```js
	build () {
		return Promise.resolve( this.resolveId( this.entry, undefined, this.resolveOptions ) )
			.then( id => this.fetchModule( id ) )
			.then( entryModule => {
				this.modules.forEach( module => {
					module.statements.forEach( statement => namespaceOpt( statement ) );
				});

				...
```

I'm not sure this is the best place to put this, but...

**Comment by Rich-Harris** [comment]
My thinking was that it'd be possible to wait to determine whether namespaces needed dynamic access until `statement.replaceIdentifiers`, in time for the namespace blocks to be [created](https://github.com/rollup/rollup/blob/master/src/Bundle.js#L223-L232), as this would avoid the need for another AST traversal, or does that create other problems?

**Comment by Victorystick** [comment]
Well... If we export a namespace, then it'll always need to be generated anyway. So we can know that this is all about creating internal namespaces or not.

The real reason not to do it in `statement.replaceIdentifiers` is that it would be called once per `bundle.render` call, resulting in an AST traversal per render. Doing it in `bundle`, before `render`, should limit the necessary traversals to a minimum.

**Comment by Rich-Harris** [comment]
But the AST traversal per render is already happening, surely? I'm just suggesting adding the optimisation to the `leave` hander of that traversal. Though glancing at the code again I realise there's probably some trickery involved in only marking the required exports during the initial pass

**Comment by Victorystick** [comment]
That's true. But I think that it's best to keep it as a separate pass, to mark only the used exports.
